src/MongoDBServiceProvider.php

```diff
@@ -66,5 +79,61 @@ public function register()
                 return new MongoConnector($this->app['db']);
             });
         });
+
+        $this->registerFlysystemAdapter();
+    }
+
+    private function registerFlysystemAdapter(): void
+    {
+        // GridFS adapter for filesystem
+        $this->app->resolving('filesystem', static function (FilesystemManager $filesystemManager) {
+            $filesystemManager->extend('gridfs', static function (Application $app, array $config) {
+                if (! class_exists(GridFSAdapter::class)) {
+                    throw new RuntimeException('GridFS adapter for Flysystem is missing. Try running "composer require league/flysystem-gridfs"');
+                }
+
+                $bucket = $config['bucket'] ?? null;
+
+                // Get the bucket from a factory function
+                if ($bucket instanceof Closure) {
+                    $bucket = $bucket($app, $config);
```

Comment:
Given the way these `if` conditions are stacked, is it possible that the Closure returns a string that is then resolved from a service, and finally resolved from a database connection? If that's not intentional, consider rewriting this to avoid this chain of resolution.

Based on the test below, I expect you may want to assert that the Closure returns a MongoDB\GridFS\Bucket instance.

Reply:
Code modified to prevent this sequence of resolution.